presto-spi/src/main/java/io/prestosql/spi/connector/ConnectorMetadata.java

```diff
@@ -602,4 +602,23 @@ default ConnectorTableProperties getTableProperties(ConnectorSession session, Co
     {
         return Optional.empty();
     }
+
+    /**
+     * Attempt to push down the provided constraint into the table. This method is provided as an alternative to
+     * {@link ConnectorMetadata#getTableLayouts(ConnectorSession, ConnectorTableHandle, Constraint, Optional)} to ease
+     * migration for the legacy API.
+     * <p>
+     * Connectors can indicate whether they don't support predicate pushdown or that the action had no effect
+     * by returning {@link Optional#empty()}. Connectors should expect this method to be called multiple times
+     * during the optimization of a given query.
+     * <p>
+     * <b>Note</b>: it's critical for connectors to return Optional.empty() if calling this method has no effect for that
+     * invocation, even if the connector generally supports pushdown. Doing otherwise can cause the optimizer
+     * to loop indefinitely.
```

Comment:
We could also check if the were no change. For example:
 - checking if the remaining predicate is empty
 - seeing that remaining predicate is equal to constraint that we are trying to push down.

Reply:
I'm not sure that is possible.  The connector could be applying the constraint internally by updating its handle, but it may not be able to enforce the change completely.  This means that there would be no external change visible, but big improvements should be happening internally.